### ISSUE TITLE
Shin Nihon Pro Wrestling Kounin - '95 Tokyo Dome Battle 7 (Japan) [bsnes v112.8]

### ISSUE THREAD

**Issue by Max833** [byuu/bsnes#214]
The screen is flickering in attract mode and before the fight in the game starts (entrance).

bsnes (fast PPU on and off), higan 106 and Mesen-S: https://streamable.com/8140g (**0:14** and **0:27**)

snes9x and hardware: https://streamable.com/48any

@James-F2: It would be very helpful if you could verify this as well. :)

![image](https://user-images.githubusercontent.com/53529402/68245912-17660b00-0018-11ea-879d-06815a8ee7e3.png)

**Comment by FitzRoyX** [comment]
Needs verified on hardware. If it is a bug, it's not a regression or something that's affected by other timing changes. Happens on every bsnes version I try all the way back to 0.51.

**Comment by James-F2** [comment]
Does NOT happen on real hardware 3Chip or 1Chip.

**Comment by Max833** [comment]
> Does NOT happen on real hardware 3Chip or 1Chip.

Thanks for verifying! :)

> Also, listen to the music in that part on real hardware in comparison to bsnes.

You mean because the attract mode music is always different? That's on hardware too.

**Comment by James-F2** [comment]
The music slows down on bsnes/mister/mesens when the screen flashing.
That does not happen on real console either.

**Comment by byuu** [comment]
Okay then, I'll look into it, thanks!

**Comment by byuu** [comment]
I imagine I'll be solving this myself like the rest, but I'll tag @SourMesen just in case.

IRQs trigger at scanlines 16, 17, 205. Both 16 and 17 enable the display, 205 disables it.

  002100=0f @ 16,1208,81
  002100=0f @ 17,1224,81
  002100=80 @ 205,1248,81

  002100=0f @ 16,1256,82
  002100=0f @ 17,1224,82
  002100=80 @ 205,1210,82

Sometimes, after the IRQ on line 16, IRQs end up not running again until the next frame.
These frames render as black.

  002100=0f @ 16,1262,83
  002100=0f @ 17,1248,84
  002100=80 @ 205,1248,84

Things recover thereafter.

  002100=0f @ 16,1210,85
  002100=0f @ 17,1232,85
  002100=80 @ 205,1212,85

  002100=0f @ 16,1206,86
  002100=0f @ 17,1228,86
  002100=80 @ 205,1202,86

At first glance, it looks like clear bad coding or a glitch to have VIRQs trigger on lines 16 and 17, and to repeat the screen enable. But the above code in and of itself is no problem: the VIRQ at line 16 enables the screen, and it stays on.

CGRAM is reloaded every frame from V=227-230, but not during our missed black frame. The white color is palette entry 255, 0x7fff. Palette entry 0 is 0x0000 (black.)

On white frames, the above layer for pixels is 0 (BG1) and below layer is also 0 (BG1.)
On black frames, the above layer is 6 (backdrop) and below layer is 0 (BG1.)

The white frame pulls the BG1 tiledata to get palette 255 to draw white.
The black frame pulls palette color 0 for the backdrop to draw black.

  //white
  00212c=15 @ 230,644,32
  00212d=05 @ 230,694,32
  IRQ @ 16,22
  002100=0f @ 16,1256,32
  00212c=00 @ 16,1292,32
  IRQ @ 17,272
  002100=0f @ 17,1224,32
  00212c=15 @ 17,1274,32
  IRQ @ 205,42
  002100=80 @ 205,1210,32
  * 0,0

  //black
  00212c=11 @ 230,648,33
  00212d=05 @ 230,698,33
  IRQ @ 16,28
  002100=0f @ 16,1262,33
  00212c=00 @ 16,1298,33 <- turns off BG1
  * 6,0
  //IRQ @ 17 doesn't fire to turn on BG1 again

  //white
  IRQ @ 17,32
  002100=0f @ 17,1248,34
  00212c=11 @ 17,1298,34
  IRQ @ 205,26
  002100=80 @ 205,1248,34
  * 0,0

So basically we need to figure out why the IRQ at line 17 sometimes doesn't fire under emulation, but always fires on real hardware.

**Comment by SourMesen** [comment]
The game does this at the end of scanline 16 to set the next irq on line 17:
![image](https://user-images.githubusercontent.com/21093973/68428172-a2124b80-0179-11ea-9325-9df9e8d6b145.png)
If the write to $4209 ends up on scanline 17, the IRQ is not triggered.

It sounds like writing to $4209/$420A to set the value to the current scanline while only V-IRQs are enabled might cause the IRQ to trigger immeditately?  e.g: like it does when you write to $4200 to turn on V-IRQs while the V timer value is the value of the current scanline.

e.g doing this after writes to $420A appears to solve the black screens:
```
if(virqEnabled && !hirqEnabled && currentScanline == verticalTimer) {
    triggerIrq();
}
```

**Comment by byuu** [comment]
I believe I've figured it out; it's really nasty. The SNES is just a nightmare of edge case games.

[shin-nihon-95.txt](https://github.com/byuu/bsnes/files/3821635/shin-nihon-95.txt)

So higan polls NMI and IRQ interrupts every four master clock cycles:

    auto CPU::step(uint clocks) -> void {
      status.irqLock = 0;
      uint ticks = clocks >> 1;
      while(ticks--) {
        counter.cpu += 2;
        tick();
        if(hcounter() & 2) pollInterrupts();
    ...

But the smallest unit of time is two master clock cycles. 6, 8, or 12 clocks per CPU cycle or DMA cycle.

The reason to poll every four is because not only should polling every two cycles not be necessary, it will actually fail my psychotic test_nmi and test_irq tests if you try: the precision of HTIME is only in dots (one dot = four clock cycles.) And NMI, plain VTIME IRQs, plain HTIME IRQs, and VTIME+HTIME IRQs all align nicely on this cour clock cycle edge.

It's convenient for us because polling interrupts is expensive.

Now if you look at the trace log, IRQs don't always trigger at exactly the same master clock cycle of course, and sometimes the IRQ triggers late enough that when it rewrites $4209=#$11 (line 17), it does so between a point where the IRQ tester sees VCOUNTER==VTIME of 16 at the first poll, and then the next test four clocks later has VCOUNTER==VTIME of 17 at the second poll. Never having them be different makes the IRQ tester think we're still effectively servicing the same interrupt and so it doesn't fire.

By polling every two clocks, we catch the edge case where VTIME is updated between these two clock cycles and VCOUNTER incrementing (technically, VCOUNTER-10 because NMI and IRQ polling is delayed from the actual V/HCOUNTER values):

    8081ad  sta $4209      [004209] A:0011 X:0000 Y:0004 S:1fa7 D:0000 B:00 nvMxdIzc 16,1346

But again, it breaks test_nmi and test_irq because the real hardware doesn't seem to test this often either.

The solution seems to be that writing to HTIME and VTIME ($4207-420a) should call the pollInterrupts() function that we call every four clock cycles immediately. The write lands right between two natural pollInterrupts() calls, and catches the 16->17 case, forcing irqHold to 0, so that a 0->1 transition can then occur and fire the interrupt.

This is one of those hellish bugs that won't show up with less accurate emulation of IRQs. Mesen-S being based on bsnes resulted in it acquiring the same bug. The same fix should work there, but again I can't guarantee this is hardware-accurate without a test ROM for my theory.

I'm not sure how we would write a test ROM to prove this, but it would be insanely challenging. And if this *isn't* the case, then I really don't know what else could allow this game to work without failing test_nmi and test_irq. I'll commit this change tomorrow, and we should go back and retest IRQ-sensitive games afterward such as Mighty Morphin Power Rangers, F1 Grand Prix, Battle Blaze, etc.

**Comment by byuu** [comment]
Oops, you posted right before I finished writing up my findings ^-^;;

Many thanks for looking at this as well.

I tested your theory out and it fails my older demo_irqtest, so unfortunately it's not the easy answer.

**Comment by SourMesen** [comment]
Ah ok, good to know there is already a test that disproves that, though.  I guess you'd need a test rom that tries to set the v timer value on every possible hclock across the scanline to validate that everything is working properly in the emulator (I'm certainly not the right person to write something like this, though)  Thankfully the bug is pretty early in the game, so I might just use the game itself as an automatic test rom on my end :p

Unrelated, but just FYI, the entire IRQ/NMI bit in Mesen-S is still a bit iffy and still fails most of the IRQ/NMI tests you wrote (+ power rangers is still broken atm). Still haven't fixed it because I am stubborn and refuse to just copy bsnes' code/logic without actually wrapping my head around what it does in terms of hardware. :p

The same problem you describe does exist though, since I'm also checking the IRQs every 4 cycles (because that's what bsnes does), and looks like I can use the same fix, too.  Thanks!

**Comment by byuu** [comment]
Passing test_nmi/test_irq is going to break your CPU core, regrettably. Its design is not compatible with my 'last cycle' testing. You'll have to do some extensive changes to make it pass, and I would recommend starting there.

I understand not wanting to 1:1 copy bsnes. From my end at least, I wouldn't care if Mesen-S were 100% identical or 100% different.

My code is based mostly on real test ROMs that I wrote, but as you can see even with 100+ edge-case tests, every single one I could think of writing back in the day, I didn't write tests that caught what Power Rangers or SNPWK '95 are doing. They do however test dozens of things I've never seen a single game rely on, like IRQs not triggering on the last dot of each frame.

(I'm really sorry I lost their source code ... that one still stings a lot. Damned flash memory.)

However, I wrote code that passed all of my tests and existing software through brute force, since I don't have a transistor map of the CPU. It doesn't mean it's the only way or the simplest way to write this code. Yet every last attempt I've ever made to simplify or change this code in any way has broken games (eg Power Rangers.) It took years of refinements and bugfixing to get it where it is.

By all means, please keep on trying to understand bsnes' IRQ logic, and once you do, maybe you can give me a small refresher ^-^;;

I would be beyond thrilled if you ended up coming up with a way to simplify the IRQ logic I use. I know it's intense but these games are just the worst as you can see. The NES has nothing on this.

Best of luck and feel free to ask me anything about SNES IRQs. And thanks for looking into this bug as well!

**Comment by byuu** [comment]
I tested all the sensitive IRQ games I could think of (F-1 Grand Prix, Battle Blaze, Mighty Morphin Power Rangers, Sink or Swim, ...) and all of my test ROMs. Things seem good.

So, here goes nothing. v112.9 pushed with the fix for this game in place. 2019 and we're still finding IRQ edge cases >_>

**Comment by James-F2** [comment]
@byuu 
Can you please share the list of the sensitive IRQ games and specific places where you 'should' see glitches with improper IRQ emulation.
That would help to test MiSTer core.
Thank you!

**Comment by Max833** [comment]
Wow, this edge case looks like a lot of work. Thank you for fixing this so quick, byyu! ^_^
